File: src/main/java/CarUnitTests.java
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class CarUnitTests {

    private Car car;

    @BeforeEach
    public void setUp() {
        // Create a Car object before each test
        car = new Car("Toyota", "Corolla", 2020, 15.0);  // 15 gallons fuel capacity
    }

    @Test
    public void testInitialFuelLevel() {
        assertEquals(0, car.getFuelLevel(), "Initial fuel level should be 0.");
    }

    @Test
    public void testRefuel() {
        car.refuel(10);
        assertEquals(10, car.getFuelLevel(), 0.01, "Fuel level should be 10 after refueling 10 gallons.");
    }

    @Test
    public void testRefuelBeyondCapacity() {
        car.refuel(20);  // Attempt to refuel more than the fuel tank capacity
        assertEquals(15, car.getFuelLevel(), 0.01, "Fuel level should not exceed the fuel capacity of 15 gallons.");
    }

    @Test
    public void testDrive() {
        car.refuel(10);
        car.drive(100);  // Driving 100 miles, which consumes 5 gallons
        assertEquals(5, car.getFuelLevel(), 0.01, "Fuel level should be 5 after driving 100 miles.");
    }

    @Test
    public void testDriveWithoutEnoughFuel() {
        car.refuel(2);
        Exception exception = assertThrows(IllegalStateException.class, () -> car.drive(100));
        assertEquals("Not enough fuel to drive the distance.", exception.getMessage());
    }

    @Test
    public void testIsEmpty() {
        assertTrue(car.isEmpty(), "The car should be empty initially.");
        car.refuel(5);
        assertFalse(car.isEmpty(), "The car should not be empty after refueling.");
    }

    @Test
    public void testNegativeRefuel() {
        Exception exception = assertThrows(IllegalArgumentException.class, () -> car.refuel(-5));
        assertEquals("Fuel amount cannot be negative.", exception.getMessage());
    }
}

File: test/CarUnitTests.cpp
// Copyright 2024 Shivam Chaturvedi
#include <gtest/gtest.h>
#include "Car.h"

class CarUnitTests : public ::testing::Test {
 protected:
    Car car;

    void SetUp() override {
        // 50 gallon capacity, 20 mpg efficiency
        car = Car("Toyota", "Corolla", 2024, 50.0, 20.0);
    }
};

TEST_F(CarUnitTests, TestInitialFuelLevel) {
    EXPECT_EQ(0, car.getRemainingRange());
}

TEST_F(CarUnitTests, TestRefuel) {
    car.refuel(10);
    // 10 gallons * 20 mpg = 200 miles
    EXPECT_EQ(200, car.getRemainingRange());
}

TEST_F(CarUnitTests, TestRefuelBeyondCapacity) {
    // Capacity is 50
    car.refuel(60);
    // 50 gallons * 20 mpg = 1000 miles
    EXPECT_EQ(1000, car.getRemainingRange());
}

TEST_F(CarUnitTests, TestDrive) {
    car.refuel(20);
    car.startEngine();
    car.drive(100);
    // (20 - 5) gallons * 20 mpg = 300 miles
    EXPECT_EQ(300, car.getRemainingRange());
}

TEST_F(CarUnitTests, TestDriveWithoutEnoughFuel) {
    car.refuel(1);
    car.startEngine();
    EXPECT_THROW(car.drive(30), std::logic_error);
}

TEST_F(CarUnitTests, TestIsEmpty) {
    EXPECT_TRUE(car.isEmpty());
    car.refuel(1);
    EXPECT_FALSE(car.isEmpty());
}

TEST_F(CarUnitTests, TestNegativeRefuel) {
    EXPECT_THROW(car.refuel(-5), std::invalid_argument);
}

TEST_F(CarUnitTests, TestDriveWithEngineOff) {
    car.refuel(20);
    EXPECT_THROW(car.drive(100), std::logic_error);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
